tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/OptimizedSlidingTimeWindowMovingAverages.java

```diff
@@ -0,0 +1,231 @@
+    public double getM15Rate() {
+        return getMinuteRate(15);
+    }
+
+    @Override
+    public double getM5Rate() {
+        return getMinuteRate(5);
+    }
+
+    @Override
+    public double getM1Rate() {
+        return getMinuteRate(1);
+    }
+
+    private double getMinuteRate(int minutes) {
+        Instant now = Instant.ofEpochSecond(0L, lastTick.get());
+        return sumBuckets(now, (int) (TimeUnit.MINUTES.toNanos(minutes) / TICK_INTERVAL));
+    }
+
+    @VisibleForTesting
+    int calculateIndexOfTick(Instant tickTime) {
+        return (int) (Duration.between(bucketBaseTime, tickTime).toNanos() / TICK_INTERVAL);
+    }
+
+    @VisibleForTesting
+    static int normalizeIndex(int index) {
+        int mod = index % NUMBER_OF_BUCKETS;
+        return mod >= 0 ? mod : mod + NUMBER_OF_BUCKETS;
+    }
+
+    private void cleanOldBuckets(Instant currentTick) {
+        int newOldestIndex;
+        Instant oldestStillNeededTime = currentTick.minus(TIME_WINDOW_DURATION).plusNanos(TICK_INTERVAL);
+        Instant youngestNotInWindow = oldestBucketTime.plus(TIME_WINDOW_DURATION);
+        if (oldestStillNeededTime.isAfter(youngestNotInWindow)) {
+            // there was no update() call for more than two whole TIME_WINDOW_DURATION
+            newOldestIndex = oldestBucketIndex;
+            oldestBucketTime = currentTick;
+        } else if (oldestStillNeededTime.isAfter(oldestBucketTime)) {
+            newOldestIndex = normalizeIndex(calculateIndexOfTick(oldestStillNeededTime));
+            oldestBucketTime = oldestStillNeededTime;
+        } else {
+            return;
+        }
+
+        cleanBucketRange(oldestBucketIndex, newOldestIndex);
+        oldestBucketIndex = newOldestIndex;
+    }
+
+    private void cleanBucketRange(int fromIndex, int toIndex) {
+        if (fromIndex < toIndex) {
+            for (int i = fromIndex; i < toIndex; i++) {
+                buckets.get(i).reset();
+            }
+        } else {
+            for (int i = fromIndex; i < NUMBER_OF_BUCKETS; i++) {
+                buckets.get(i).reset();
+            }
+            for (int i = 0; i < toIndex; i++) {
+                buckets.get(i).reset();
+            }
+        }
+    }
+
+    private long sumBuckets(Instant toTime, int numberOfBuckets) {
+
+        // increment toIndex to include the current bucket into the sum
+        int toIndex = normalizeIndex(calculateIndexOfTick(toTime) + 1);
+        int fromIndex = normalizeIndex(toIndex - numberOfBuckets);
+
+        long sum = 0;
+        if (fromIndex < toIndex) {
+            for (int i = fromIndex; i < toIndex; i++) {
+                sum += buckets.get(i).longValue();
+            }
+        } else {
+            for (int i = fromIndex; i < NUMBER_OF_BUCKETS; i++) {
+                sum += buckets.get(i).longValue();
+            }
+            for (int i = 0; i < toIndex; i++) {
+                sum += buckets.get(i).longValue();
+            }
+        }
+        return sum;
```

Comment:
these are the optimized bits from https://github.com/dropwizard/metrics/pull/4936